### DISCUSSION TITLE
Direct input using codepoint like MS-IME (codepoint + F5)

### DISCUSSION THREAD

**Discussion by tats-u** [google/mozc#875] [Ideas]
MS-IME supports easier-to-input direct input via codepoint.

For example, MS-IME can input "𠮟" (U+20B9F) by:

1. Turn IME on if off
2. Input ２→０→ｂ→９→ｆ
3. Press <kbd>F5</kbd>

MS-IME also can convert "`００**あ**５" into "¥"(U+00A5) by <kbd>F5</kbd>.

This method takes less number of key types to input characters difficult to type in other ways.

**Comment by yukawa** [comment]
Thank you for letting us know about this compatibility gap.

I believe the feature itself is implementable in Mozc just by following how other existing commands such as `ConvertToHalfAlphanumeric` and `ConvertToHiragana` have been implemented.
https://github.com/google/mozc/blob/c2fcbf6515c5884437977de46187c16a8cb7bb50/src/data/keymap/ms-ime.tsv?plain=1#L36-L41

https://github.com/search?q=repo%3Agoogle%2Fmozc%20ConvertToHalfAlphanumeric%20&type=code

Those who design and implement such a new command (`ConvertFromCodePoint`?) probably end up having to touch multiple files, but overall the task should be straightforward and can be done in a test-driven manner.

While we cannot promise whether this feature will be implemented or not in the future, given that this is a legitimate compatibility gap between MS-IME and Mozc, I guess we would be happy to have this feature if someone implemented it with sufficient tests.

----

In the mean time you can achieve a similar result as follows.
 1. Activate Mozc
 2. Type <kbd>U</kbd>, <kbd>+</kbd>, <kbd>2</kbd>, <kbd>0</kbd>, <kbd>b</kbd>, <kbd>9</kbd>, <kbd>f</kbd>
 3. Press <kbd>Space</kbd> or <kbd>Tab</kbd>
<img width="270" alt="image" src="https://github.com/google/mozc/assets/5861397/06094e27-aa2d-4016-8fb3-854c261a676d">

This feature is implemented in [`src/rewriter/unicode_rewriter.cc`](https://github.com/google/mozc/blob/c2fcbf6515c5884437977de46187c16a8cb7bb50/src/rewriter/unicode_rewriter.cc).

**Reply by tats-u** [reply]
@yukawa 

It look like it has not a high priority. It would not so bad to implement it by myself when I can preserve a time.

> In the mean time you can achieve a similar result as follows.

I know it but it requires more number of types and the hold of <kbd>Shift</kbd>.

**Reply by tats-u** [reply]
Is it a bad ~~idea~~ design to reuse `Session::ConvertToTransliteration` and modify `SessionConverter::ConvertToTransliteration` by adding the followindg code?

src\session\session.cc

```diff
 bool Session::ConvertToTransliteration(
     commands::Command *command,
     const transliteration::TransliterationType type) {
   if (!(context_->state() &
         (ImeContext::CONVERSION | ImeContext::COMPOSITION))) {
     return DoNothing(command);
   }
   command->mutable_output()->set_consumed(true);
 
   if (!context_->mutable_converter()->ConvertToTransliteration(
           context_->composer(), type)) {
     return false;
   }
   SetSessionState(ImeContext::CONVERSION, context_.get());
   Output(command);
   return true;
 }
 
+bool Session::TryConvertFromCodePoint(commands::Command *command) {
+  return ConvertToTransliteration(command, transliteration::CODEPOINT);
+}
+
 bool Session::ConvertToHiragana(commands::Command *command) {
   return ConvertToTransliteration(command, transliteration::HIRAGANA);
 }
 
 bool Session::ConvertToFullKatakana(commands::Command *command) {
   return ConvertToTransliteration(command, transliteration::FULL_KATAKANA);
 }
```

It depends on the following feature, doesn't it?

![image](https://github.com/user-attachments/assets/53221abd-a809-4965-b411-c69e044964d1)

**Reply by yukawa** [reply]
> Is it a bad idea to reuse `Session::ConvertToTransliteration` and modify `SessionConverter::ConvertToTransliteration` by adding the followindg code?

@hiroyuki-komatsu Do you have any concern?

**Reply by tats-u** [reply]
It looks like the following array corresponds to the order of the above candidates:

src\transliteration\transliteration.h

```diff
 static const TransliterationType TransliterationTypeArray[NUM_T13N_TYPES] = {
     HIRAGANA,         FULL_KATAKANA,          HALF_ASCII,    HALF_ASCII_UPPER,
     HALF_ASCII_LOWER, HALF_ASCII_CAPITALIZED, FULL_ASCII,    FULL_ASCII_UPPER,
     FULL_ASCII_LOWER, FULL_ASCII_CAPITALIZED, HALF_KATAKANA,
+    FULL_ASCII_LOWER, FULL_ASCII_CAPITALIZED, HALF_KATAKANA, CODEPOINT,
 };
```

It looks like we should not add a new entry there, or a codepoint candidate will be always appended to the candidates of usual conversions.
I wonder if we should do the following instead:

```diff
diff --git a/src/transliteration/transliteration.h b/src/transliteration/transliteration.h
index 89eda27dc..840791ee9 100644
--- a/src/transliteration/transliteration.h
+++ b/src/transliteration/transliteration.h
@@ -62,11 +62,13 @@ enum TransliterationType {
   FULL_ASCII_CAPITALIZED,
   // "ｶﾀｶﾅ"
   HALF_KATAKANA,
+  NUM_PUBLIC_T13N_TYPES,
+  CODEPOINT = NUM_PUBLIC_T13N_TYPES,
   NUM_T13N_TYPES
 };
 typedef std::vector<std::string> Transliterations;

-static const TransliterationType TransliterationTypeArray[NUM_T13N_TYPES] = {
+static const TransliterationType TransliterationTypeArray[NUM_PUBLIC_T13N_TYPES] = {
     HIRAGANA,         FULL_KATAKANA,          HALF_ASCII,    HALF_ASCII_UPPER,
     HALF_ASCII_LOWER, HALF_ASCII_CAPITALIZED, FULL_ASCII,    FULL_ASCII_UPPER,
     FULL_ASCII_LOWER, FULL_ASCII_CAPITALIZED, HALF_KATAKANA,
```

**Reply by hiroyuki-komatsu** [reply]
> > Is it a bad idea to reuse `Session::ConvertToTransliteration` and modify `SessionConverter::ConvertToTransliteration` by adding the followindg code?
> 
> @hiroyuki-komatsu Do you have any concern?

It depends on the expected behavior. Could you please confirm what happens on F5 with the following cases?

* The composition text is not a code point (e.g. "xyz")
* The composition text is a control code (e.g. "0" or "FFEE")
